cluster/router/condition/listenable_router.go

```diff
@@ -0,0 +1,149 @@
+
+import (
+	"fmt"
+)
+
+import (
+	perrors "github.com/pkg/errors"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/config"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/config_center"
+	"github.com/apache/dubbo-go/protocol"
+	"github.com/apache/dubbo-go/remoting"
+)
+
+const (
+	// Default priority for listenable router, use the maximum int64 value
+	listenableRouterDefaultPriority = ^int64(0)
+)
+
+// listenableRouter Abstract router which listens to dynamic configuration
+type listenableRouter struct {
+	conditionRouters []*ConditionRouter
+	routerRule       *RouterRule
+	url              *common.URL
+	force            bool
+	priority         int64
+}
+
+// RouterRule Get RouterRule instance from listenableRouter
+func (l *listenableRouter) RouterRule() *RouterRule {
+	return l.routerRule
+}
+
+func newListenableRouter(url *common.URL, ruleKey string) (*AppRouter, error) {
+	if ruleKey == "" {
+		return nil, perrors.Errorf("NewListenableRouter ruleKey is nil, can't create Listenable router")
+	}
+	l := &AppRouter{}
+
+	l.url = url
+	l.priority = listenableRouterDefaultPriority
+
+	routerKey := ruleKey + constant.ConditionRouterRuleSuffix
+	//add listener
+	dynamicConfiguration := config.GetEnvInstance().GetDynamicConfiguration()
+	if dynamicConfiguration == nil {
+		return nil, perrors.Errorf("Get dynamicConfiguration fail, dynamicConfiguration is nil, init config center plugin please")
+	}
+
+	dynamicConfiguration.AddListener(routerKey, l)
+	//get rule
+	rule, err := dynamicConfiguration.GetRule(routerKey, config_center.WithGroup(config_center.DEFAULT_GROUP))
+	if len(rule) == 0 || err != nil {
+		return nil, perrors.Errorf("Get rule fail, config rule{%s},  error{%v}", rule, err)
+	}
+	l.Process(&config_center.ConfigChangeEvent{
+		Key:        routerKey,
+		Value:      rule,
+		ConfigType: remoting.EventTypeUpdate})
+
+	logger.Info("Init app router success")
+	return l, nil
+}
+
+// Process Process config change event , generate routers and set them to the listenableRouter instance
+func (l *listenableRouter) Process(event *config_center.ConfigChangeEvent) {
+	logger.Infof("Notification of condition rule, change type is:[%s] , raw rule is:[%v]", event.ConfigType, event.Value)
+	if remoting.EventTypeDel == event.ConfigType {
+		l.routerRule = nil
+		l.conditionRouters = make([]*ConditionRouter, 0)
+		return
+	}
+	content, ok := event.Value.(string)
+	if !ok {
+		msg := fmt.Sprintf("Convert event content fail,raw content:[%s] ", event.Value)
+		logger.Error(msg)
+		return
+	}
+
+	routerRule, err := Parse(content)
+	if err != nil {
+		logger.Errorf("Parse condition router rule fail,error:[%s] ", err)
+		return
+	}
+	l.generateConditions(routerRule)
+}
+
+func (l *listenableRouter) generateConditions(rule *RouterRule) {
+	if rule == nil || !rule.Valid {
+		return
+	}
+	l.conditionRouters = make([]*ConditionRouter, 0)
```

Comment:
l.conditionRouters = make([]*ConditionRouter, 0, len(l.routers))

Reply:
should be ```l.conditionRouters = make([]*ConditionRouter, 0, len(rule.Conditions))```